tests/unit/check_terminal.c

```diff
@@ -0,0 +1,266 @@
+{
+    setup();
+    teardown();
+}
+END_TEST
+
+// This test fails for the following chars
+// Sent - Read
+// (0x00..0x1d) - 0x1d (ASCII Group Separator?)
+// 0x7e - 0x80 (~ and non-ascii)
+// 0x7f - 0x80 (DEL & non-ascii)
+
+// If canonical mode and signals are disabled (in terminal.c)
+// 0x0a00 - 0x0d00 (CR & LF switched)
+// 0x0d00 - 0x0a00
+// 0x1100 - 0x1200
+START_TEST(test_event_handler)
+{
+    osd_result rv;
+
+    const uint16_t START = 0x80, END = 0x100;
+
+    struct osd_packet *packet;
+    osd_packet_new(&packet, osd_packet_sizeconv_payload2data(1));
+    osd_packet_set_header(packet, mock_hostmod_diaddr, mock_dem_uart_diaddr,
+                          OSD_PACKET_TYPE_EVENT, EV_LAST);
+
+    for (uint16_t c = START; c < END; ++c) {
+        packet->data.payload[0] = c;
+        rv = mock_host_controller_queue_data_packet(packet);
+        ck_assert_int_eq(rv, OSD_OK);
+    }
+
+    // This is only needed when canonical mode is enabled (in terminal.c)
+    //packet->data.payload[0] = '\n';
+    //rv = mock_host_controller_queue_data_packet(packet);
+    //ck_assert_int_eq(rv, OSD_OK);
+
+    mock_host_controller_wait_for_event_tx();
+    osd_packet_free(&packet);
+
+    int pts = open(osd_terminal_get_pts_path(terminal_ctx), O_RDONLY | O_NOCTTY);
+    if (pts < 0) {
+        perror("open()");
+        return;
+    }
+
+    int ret;
+    uint8_t c;
+    for (uint16_t exp = START; exp < END; ++exp) {
+        ret = read(pts, &c, 1);
+        if (ret <= 0) {
+            perror("read()");
+        }
+
+        fprintf(stderr, "c: 0x%x, exp: 0x%x\n", c, exp);
+        ck_assert_int_eq(c, exp);
+    }
+}
+END_TEST
+
+// This test fails for the following chars:
+// Expected - Received (Little Endian)
+// 0xa00 - 0xd00 (LF turned into CR)
+START_TEST(test_terminal_thread)
+{
+    osd_result rv;
+
+    const uint16_t START = 0xb, END = 0x0100;
+    assert(END > START);
+
+    struct osd_packet *packet;
+    osd_packet_new(&packet, osd_packet_sizeconv_payload2data(1));
+    osd_packet_set_header(packet, mock_dem_uart_diaddr, mock_hostmod_diaddr,
+                          OSD_PACKET_TYPE_EVENT, EV_LAST);
+
+    uint8_t buf[END - START];
+    for (uint16_t c = START; c < END; ++c) {
+        packet->data.payload[0] = c;
+        mock_host_controller_expect_data_req(packet, NULL);
+
+        buf[c - START] = c;
+    }
+
+    int pts = open(osd_terminal_get_pts_path(terminal_ctx), O_WRONLY | O_NOCTTY);
+    if (pts < 0) {
+        perror("open()");
+        return;
+    }
+
+    for (int written = 0, ret; written < sizeof(buf); /*NOP*/) {
+        ret = write(pts, buf + written, sizeof(buf) - written);
+        if (ret <= 0) {
+            perror("write()");
+        }
+
+        written += ret;
+    }
+
+    mock_host_controller_wait_for_requests();
+
+    osd_packet_free(&packet);
+}
+END_TEST
+
+START_TEST(test_file_permissions)
+{
+    struct stat stats;
+
+    if(stat(osd_terminal_get_pts_path(terminal_ctx), &stats) == -1) {
```

Comment:
That's three calls in one, which makes it hard to see where the failure comes from.
```c
const char* pts_path = osd_terminal_get_pts_path(terminal_ctx);
ck_assert(pts_path);
int irv = stat(pts_path, &pts_stat);
ck_assert(irv == 0);
```

I would probably leave the checks at that level, checking beyond that is easy to get wrong and does not really help much. (I.e. don't check for the "right" permissions.)